sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -401,15 +401,30 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
                     |the ${ordinalNumber(ti + 1)} table has ${child.output.length} columns
                   """.stripMargin.replace("\n", " ").trim())
               }
+              val isUnion = operator.isInstanceOf[Union]
+              val dataTypesAreCompatibleFn = if (isUnion) {
+                // `TypeCoercion` takes care of type coercion already. If any columns or nested
+                // columns are not compatible, we detect it here and throw analysis exception.
+                val typeChecker = (dt1: DataType, dt2: DataType) => {
+                  !TypeCoercion.findWiderTypeForTwo(dt1.asNullable, dt2.asNullable).isEmpty
```

Comment:
I know it's from the old code. But is it necessary? The analyzer can add implicit casts to make the types the same.

Reply:
It is not always able to cast the types between children of union. For incompatible types, we need to find it out and throw analysis error here. Do I misunderstand it?

Reply:
For compatible types, spark will add implicit cast and the types should be equal. So I think a simple `dt1 == dt2` should be sufficient here?

Reply:
Oh, I spent a little time to recall why I keep original check logic.

It is because if `TypeCoercion` fails to find compatible types for all columns, it won't add cast for all. It is all or nothing logic here.

So if we only check `dt1 == dt2` here, we compare the original data types even some of them are compatible.

`AnalysisErrorSuite` has one example. One relation has `short, string, double, decimal`, another one has `string, string, string, map`.

The first three columns are compatible, only the fourth isn't. So `TypeCoercion` fails to add casts for all.

If we compare `dt1 == dt2`, the error will be like "short is not compatible with string". But currently we get like "decimal is not compatible with map".

Reply:
I see, thanks for the explanation!

I think it's not a good practice to rely on the code logic that is far away from the current code. Is it possible to update the type coercion rule and add implicit casts using best efforts?

Reply:
You mean to add cast for compatible ones, and ignore these cannot? let me try if it's possible.

Reply:
> You mean to add cast for compatible ones, and ignore these cannot?

Yes, by doing that we can just do `dt1 == dt2` here, and makes the code less coupled.

Reply:
Ok, it works. I changed `TypeCoercion` to follow the discussed behavior.